cookbook/case_studies/feature_engineering/sqlite_datacleaning/workflow.py

```diff
@@ -0,0 +1,77 @@
+import typing
+
+import pandas as pd
+from flytekit import Workflow, kwtypes, reference_task
+from flytekit.extras.sqlite3.task import SQLite3Config, SQLite3Task
+from flytekit.types.file import FlyteFile
+from flytekit.types.schema import FlyteSchema
+
+
+@reference_task(
```

Comment:
The reference tasks aren't working when included in the `case_studies` directory (works in `core`). 
The error that looks like: `Failed to pull image "flytecookbook:sqlite_datacleaning-aedbd6fe44051c171fd966c280c5c3036f658831": rpc error: code = Unknown desc = Error response from daemon: pull access denied for flytecookbook, repository does not exist or may require 'docker login': denied: requested access to the resource is denied`. 
How do I resolve this?

Reply:
that image looks wrong? did you push the image? how did you run this?

Reply:
I ran `SANDBOX=1 make -C cookbook/case_studies/feature_engineering/sqlite_datacleaning/ register` command. I also see the image on docker desktop.

Reply:
I see the image on docker desktop as well.

Reply:
```bash
$ docker run flytecookbook:sqlite_datacleaning-0774958b4ebefa12d909ee1eb7cb10de0c1790fa 
$ docker pull flytecookbook:sqlite_datacleaning-0774958b4ebefa12d909ee1eb7cb10de0c1790fa
Error response from daemon: pull access denied for flytecookbook, repository does not exist or may require 'docker login': denied: requested access to the resource is denied
```
`docker run` is working fine whereas `docker pull` is resulting in this error.

Reply:
If you build an image locally, you will be able to access it, but when you `pull`, docker will actually try to make a network connection to the docker registry and download it from there.  That is why it works when you run it but not when you pull it.

I can do the following:
```
ytong@argus:~ [arn:aws:eks:us-west-2:311464980720:cluster/opta-hosted-testorg12]$ mkdir tst
ytong@argus:~ [arn:aws:eks:us-west-2:311464980720:cluster/opta-hosted-testorg12]$ cd tst
ytong@argus:~/tst [arn:aws:eks:us-west-2:311464980720:cluster/opta-hosted-testorg12]$ cat > Dockerfile
FROM ubuntu:focal
^C
ytong@argus:~/tst [arn:aws:eks:us-west-2:311464980720:cluster/opta-hosted-testorg12]$ docker build -t abc.com/dummy:123 .
[+] Building 0.1s (5/5) FINISHED
 => [internal] load build definition from Dockerfile                                                                                                                            0.0s
 => => transferring dockerfile: 60B                                                                                                                                             0.0s
 => [internal] load .dockerignore                                                                                                                                               0.0s
 => => transferring context: 2B                                                                                                                                                 0.0s
 => [internal] load metadata for docker.io/library/ubuntu:focal                                                                                                                 0.0s
 => CACHED [1/1] FROM docker.io/library/ubuntu:focal                                                                                                                            0.0s
 => exporting to image                                                                                                                                                          0.0s
 => => exporting layers                                                                                                                                                         0.0s
 => => writing image sha256:f2bed4af2163116084464018b01076094f252d4c7fee7bad02f8e73ef7a4e78d                                                                                    0.0s
 => => naming to abc.com/dummy:123                                                                                                                                              0.0s
ytong@argus:~/tst [arn:aws:eks:us-west-2:311464980720:cluster/opta-hosted-testorg12]$ docker run abc.com/dummy:123 echo hi
hi
```

I can run and create an image under abc.com even though I clearly don't own that domain.

If you don't specify a domain, one is specified by default for you.  If I just run `docker build -t nodomain:123 .` the fully qualified name is actually `docker.io/library/nodomain:123`.  In the above example, in the `FROM`, `ubuntu:focal` is actually
```
$ docker pull docker.io/library/ubuntu:focal
focal: Pulling from library/ubuntu
Digest: sha256:adf73ca014822ad8237623d388cedf4d5346aa72c270c5acc01431cc93e18e2d
Status: Image is up to date for ubuntu:focal
docker.io/library/ubuntu:focal
```

All this is to say that the reference task that you're using wasn't created correctly.  If you were to look at the JSON representation for the underlying task in the UI, the image would just be `flytecookbook:sqlite_datacleaning-aedbd6fe44051c171fd966c280c5c3036f658831` whereas it should be `ghcr.io/flyteorg/flytecookbook:sqlite_datacleaning-aedbd6fe44051c171fd966c280c5c3036f658831` (assuming the rest of the name is correct).  Without that first part, it's trying to pull the image from dockerhub (which runs `docker.io/library`) instead of github.

Reply:
Thank you, @wild-endeavor! Where to "add" ghcr.io while referencing the actual `flytecookbook:sqlite_datacleaning` image for the case_studies repo? The referencing is happening while using the `@reference_task` decorator, and I believe this all happens internally (I ain't mentioning it anywhere, do I have to specifically say that `ghcr.io` is where it has to pull from, if so how?)
Moreover, how's it working for `flytecookbook:core`? Isn't the config the same in case_studies as well? How is `flytecookbook:core` image getting pulled automatically (even when `ghcr.io` isn't mentioned anywhere)?

Reply:
The issue's resolved after I've used `ECR` to upload the image. Thank you, @kumare3 and @wild-endeavor.